src/Cornelis/Offsets.hs

```diff
@@ -1,27 +1,139 @@
 {-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingStrategies #-}
 
 module Cornelis.Offsets
-  ( module Cornelis.Offsets
-  , LineNumber(..)
+  ( Indexing(..)
+  , Unit(..)
+  , Index()
   , Offset(..)
-  , BufferOffset
-  , LineOffset
+  , Pos(..)
+  , Interval(..)
+  , LineNumber
+  , AgdaIndex
+  , AgdaOffset
+  , AgdaPos
+  , AgdaInterval
+  , VimIndex
+  , VimOffset
+  , VimPos
+  , VimInterval
+  , zeroIndexed
+  , oneIndexed
+  , fromZeroIndexed
+  , fromOneIndexed
+  , zeroIndex
+  , oneIndex
+  , incIndex
+  , (.+)
+  , (.-.)
+  , offsetPlus
+  , textToBytes
+  , charToBytes
+  , toBytes
+  , fromBytes
+  , containsPoint
+  , addCol
   ) where
 
-import           Cornelis.Types
-import           Cornelis.Types.Agda
+import           Data.Aeson (FromJSON)
 import qualified Data.ByteString as BS
 import           Data.Coerce (coerce)
-import           Data.Int
 import qualified Data.Text as T
 import           Data.Text.Encoding (encodeUtf8)
+import           GHC.Generics (Generic)
+import           GHC.Stack (HasCallStack)
+import           Prettyprinter (Pretty)
 
+-- TODO: Distinguish offsets and positions; offsets have arithmetic, positions don't
+
+-- | Indexing scheme: whether the first index is zero or one.
+data Indexing = OneIndexed | ZeroIndexed
+
+-- | What are we counting?
+data Unit = Byte | CodePoint | Line
+
+-- | The constructor is hidden, use 'zeroIndexed' and 'oneIndexed' to construct it,
+-- and 'fromZeroIndexed' and 'fromOneIndexed' to destruct it.
+newtype Index (e :: Unit) (i :: Indexing) = Index Int
+  deriving newtype (Eq, Ord, Show, Read, FromJSON, Pretty)
+
+-- | It doesn't seem worth the trouble to hide this constructor.
+newtype Offset (e :: Unit) = Offset Int
+  deriving newtype (Eq, Ord, Show, Read, FromJSON, Pretty)
+
+-- | Position in a text file as line-column numbers. This type is indexed by
+-- the units of the columns (@Byte@ or @CodePoint@) and by the indexing scheme
+-- of lines and columns.
+data Pos e i j = Pos
+  { p_line :: Index 'Line i
+  , p_col :: Index e j
+  } deriving (Eq, Ord, Show, Generic)
+
+data Interval p = Interval { iStart, iEnd :: !p }
+  deriving (Eq, Ord, Show, Functor, Foldable, Traversable, Generic)
+
+-- Common specializations
+
+type LineNumber = Index 'Line
+
+type AgdaIndex = Index 'CodePoint 'OneIndexed
+type AgdaOffset = Offset 'CodePoint
+type AgdaPos = Pos 'CodePoint 'OneIndexed 'OneIndexed
+type AgdaInterval = Interval AgdaPos
+
+type VimIndex = Index 'Byte
+type VimOffset = Offset 'Byte
+type VimPos = Pos 'Byte 'ZeroIndexed 'ZeroIndexed
+type VimInterval = Interval VimPos
+
+-- To pass indices to and from external sources.
+
+-- | Mark a raw index as zero-indexed.
+zeroIndexed :: Integral a => a -> Index e 'ZeroIndexed
```

Comment:
I don't have a great naming suggestion here, but I get this confused with `zeroIndex`. Maybe `toZeroIndexed` here for symmetry with `fromZeroIndexed`?